db/view/view.cc

```diff
@@ -1907,13 +1907,18 @@ future<> view_update_generator::mutate_MV(
                 if (*target_endpoint == *remote_it) {
                     remote_endpoints.erase(remote_it);
                 } else {
-                    std::swap(*target_endpoint, *remote_it);
+                    auto target_remote_it = std::find(remote_endpoints.begin(), remote_endpoints.end(), *target_endpoint);
+                    if (target_remote_it != remote_endpoints.end()) {
+                        target_endpoint = *remote_it;
+                        remote_endpoints.erase(remote_it);
+                    } else {
+                        std::swap(*target_endpoint, *remote_it);
+                    }
                 }
             }
-        }
-        // It's still possible that a target endpoint is duplicated in the remote endpoints list,
-        // so let's get rid of the duplicate if it exists
-        if (target_endpoint) {
+        } else if (target_endpoint) {
+            // It's still possible that a target endpoint is duplicated in the remote endpoints list,
+            // so let's get rid of the duplicate if it exists
```

Comment:
Actually, I'm not even sure what is this `if` even for. It looks like we're looking for the `target_endpoint` which can't be a pending endpoint (we dont pair with pending endpoints) in `remote_endpoints` which are pending endpoints...
Same with `*target_endpoint == *remote_it` above

@nyh do you remember what it was for?

Reply:
The original code was sub-optimal in theory in case both my_address and *target_endpoint are in remote_endpoints.  In this case swap would keep 2 instances of *target_endpoint and the find below wouldn't find it, as you mentioned above.
But this is all theoretical since I don't think we may have more than one pending endpoint.

Reply:
> Actually, I'm not even sure what is this `if` even for. It looks like we're looking for the `target_endpoint` which can't be a pending endpoint (we dont pair with pending endpoints) in `remote_endpoints` which are pending endpoints... Same with `*target_endpoint == *remote_it` above
> 
> @nyh do you remember what it was for?

Since we're sending the view update to target_endpoint plus to all remote_endpoints, the exercise above wants to make sure that if my_address is in remote_endpoints then it is moved into or swapped with target_endpoint.
Since the local view update optimization is done only for target_endpoint (see `if (target_endpoint && *target_endpoint == my_address)`)
If my_address is not in remote_endpoints, then we want to erase *target_address from remote_endpoints so that `apply_to_remote_endpoints` would send it only once.

Reply:
> If my_address is not in remote_endpoints, then we want to erase *target_address from remote_endpoints so that `apply_to_remote_endpoints` would send it only once.

This seems to be the intention, but can `*target_endpoint ` be really duplicated in `remote_endpoints`? `remote_endpoints` are the pending endpoints (at least initially) and `*target_endpoint` is the endpoint we paired with. We don't pair with pending endpoints (`get_view_natural_endpoint` uses the `view_erm->get_replicas()` list to select the endpoint) so `*target_endpoint` shouldn't be in `remote_endpoints`.
Maybe it was different in the past, but now, instead of handling these cases, perhaps it would be better to throw an error on them to detect when something unexpected happens.

Also, with tablets it seems in fact impossible to have more than one pending endpoint, but I think it can happen for vnodes.

Reply:
> > If my_address is not in remote_endpoints, then we want to erase *target_address from remote_endpoints so that `apply_to_remote_endpoints` would send it only once.
> 
> This seems to be the intention, but can `*target_endpoint ` be really duplicated in `remote_endpoints`? `remote_endpoints` are the pending endpoints (at least initially) and `*target_endpoint` is the endpoint we paired with. We don't pair with pending endpoints (`get_view_natural_endpoint` uses the `view_erm->get_replicas()` list to select the endpoint) so `*target_endpoint` shouldn't be in `remote_endpoints`. Maybe it was different in the past, but now, instead of handling these cases, perhaps it would be better to throw an error on them to detect when something unexpected happens.

Let's leave that for a follow up and keep the focus in this PR on rack-aware pairing.

> Also, with tablets it seems in fact impossible to have more than one pending endpoint, but I think it can happen for vnodes.

How can it happen with vnodes?
token_metadata supports only 1 outstanding topology node operation.

Reply:
> Let's leave that for a follow up and keep the focus in this PR on rack-aware pairing.
> 
> > Also, with tablets it seems in fact impossible to have more than one pending endpoint, but I think it can happen for vnodes.
> 
> How can it happen with vnodes? token_metadata supports only 1 outstanding topology node operation.

It seems I was a bit confused about what happens to the token ring during a topology operation. You're right, there should be at most 1 pending replica even with vnodes